### DISCUSSION TITLE
Node version error message

### DISCUSSION THREAD

**Discussion by davegardner** [11ty/eleventy#2264] [Q&A] [answered]
I start a terminal (bash/Ubutu), cd to the project directory and run 

`node -v `
`v16.14.0`
`npx @11ty/eleventy`
`Eleventy requires Node 12. You will need to upgrade Node to use Eleventy!`

If I start a terminal from within VS Code it runs just fine.

Any ideas about what I've screwed up?

Things I've tried:

- restarting the machine
- reinstalling node
- reinstalling eleventy

**Comment by pdehaan** [comment]
Hhmm, that's interesting. I'm not an Ubuntu user, but I can see if I can install a VM tomorrow on macOS via Parallels and see if I can reproduce that. But still odd that your terminal and VS Code would give different results, unless they're using slightly different shells or possibly different $PATH values or something.

https://flaviocopes.com/node-version-runtime/ looks promising. Might want to create a quick script that dumps the `process.versions` variable and confirm in both environments.

```js
// ver-check.js
console.log(process.versions);
```

```sh
node ver-check.js
{
  node: '17.4.0',
  v8: '9.6.180.15-node.13',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '102',
  nghttp2: '1.45.1',
  napi: '8',
  llhttp: '6.0.4',
  openssl: '3.0.1+quic',
  cldr: '40.0',
  icu: '70.1',
  tz: '2021a3',
  unicode: '14.0',
  ngtcp2: '0.1.0-DEV',
  nghttp3: '0.1.0-DEV'
}
```

I seem to get consistent results on macOS 12.2 via my iTerm, and VS Code's terminal, and a package.json script that runs my <kbd>node ver-check</kbd> script.

**Reply by davegardner** [reply]
Sorry for late reply but I was out of contact for a few days. Tried your suggestion and found nothing surprising.

Path is 

`bash:/home/dave/dotnet:/home/dave/.netlify/helper/bin:/home/dave/.nvm/versions/node/v16.14.0/bin:/home/dave/dotnet:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/dave/Projects/flutter/bin:/home/dave/Projects/flutter/bin:`

Versions.js dumps:

   dave@nuc3:~/Shared/Projects/SevenSeas$ node versions
{
  node: '16.14.0',
  v8: '9.4.146.24-node.20',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '93',
  nghttp2: '1.45.1',
  napi: '8',
  llhttp: '6.0.4',
  openssl: '1.1.1m+quic',
  cldr: '40.0',
  icu: '70.1',
  tz: '2021a3',
  unicode: '14.0',
  ngtcp2: '0.1.0-DEV',
  nghttp3: '0.1.0-DEV'
}

**Reply by pdehaan** [reply]
@davegardner But are those results consistent between your Bash terminal and your VSCode terminal?
But thanks for the reminder. I did install Ubuntu VM in my parallels the other day for work, but still need to figure out how to best install Node and VSCode so I can see if I can reproduce this locally.

**Reply by davegardner** [reply]
Yes. Same in VS Code terminal and system bash terminal.

**Reply by davegardner** [reply]
I am considering rebuilding this machine, but inconsistencies like this annoy me and I could not think of an explanation beyond what you have suggested (environment/versions). So although it is a practical problem I expect a rebuild will fix it.

**Comment by davegardner** [comment]
I strongly suspect this has something to do with snap. Both node and code are installed via snap. How does that work with nvm?

**Reply by pdehaan** [reply]
OK, you motivated me to _actually_ finish installing Ubuntu 20.0x so I can use VS Code and Node in a VM (macOS 12.2 on M1 Mac via Parallels 17.something).

```sh
sudo apt install curl
curl https://raw.githubusercontent.com/creationix/nvm/master/install.sh | bash
source .bashrc

nvm i 17
node --version # v17.7.1
```

I tried using whatever this is, but it didn't seem to like it:
```sh
sudo snap install --classic code
error: snap "code" is not available on stable for this architecture (arm64) but exists on other
       architectures (amd64).
```

Then downloaded VS Code for Ubuntu via Firefox and got this:
```sh
sudo dpkg -i code_1.65.0-1646220682_amd64.deb 
dpkg: error processing archive code_1.65.0-1646220682_amd64.deb (--install):
 package architecture (amd64) does not match system (arm64)
Errors were encountered while processing:
 code_1.65.0-1646220682_amd64.deb
```

Then randomly guessed I might need an arm64 build since I'm on an Apple M1 Max chip and downloaded a new new build and tried this:
```sh
sudo dpkg -i code_1.65.2-1646922911_arm64.deb 
Selecting previously unselected package code.
(Reading database ... 216814 files and directories currently installed.)
Preparing to unpack code_1.65.2-1646922911_arm64.deb ...
Unpacking code (1.65.2-1646922911) ...
dpkg: dependency problems prevent configuration of code:
 code depends on libxss1; however:
  Package libxss1 is not installed.

dpkg: error processing package code (--install):
 dependency problems - leaving unconfigured
Processing triggers for gnome-menus (3.36.0-1ubuntu1) ...
Processing triggers for desktop-file-utils (0.24-1ubuntu3) ...
Processing triggers for mime-support (3.64ubuntu1) ...
Processing triggers for shared-mime-info (1.15-1) ...
Errors were encountered while processing:
 code
```

"The Internet" then told me to try this:
```sh
sudo apt install -f
```

And by some small miracle, that seemed to get <kbd>code .</kbd> working for me on my Ubuntu 20.04.2 ARM64 Parallels VM.
I can now build my basic "hello world" Eleventy site via the Terminal:
```sh
npm run build

> 11ty-test@1.0.0 build
> eleventy

[11ty] Writing www/index.html from ./src/index.njk
[11ty] Wrote 1 file in 0.04 seconds (v1.0.0)
```

Or by opening a Terminal within VS Code and running <kbd>npm run build</kbd> from within my project directory.

**Reply by pdehaan** [reply]
Disclaimer:
![](https://starecat.com/content/wp-content/uploads/dog-i-have-no-idea-what-im-doing-science-chemistry.jpg)

**Reply by davegardner** [reply]
I'm the one with the safety goggles alright.
I was wrong. Node was originally installed via apt, then I installed nvm. It has not ever been installed via snap. So:

```
dave@nuc3:~/snap$ snap list
Name                     Version                     Rev    Tracking         Publisher     Notes
bare                     1.0                         5      latest/stable    canonical✓    base
code                     c722ca6c                    91     latest/stable    vscode✓       classic
core                     16-2.54.4                   12821  latest/stable    canonical✓    core
core18                   20211215                    2284   latest/stable    canonical✓    base
core20                   20220304                    1376   latest/stable    canonical✓    base
gimp                     2.10.28                     383    latest/stable    snapcrafters  -
gitkraken                8.2.1                       188    latest/stable    gitkraken✓    classic
gnome-3-28-1804          3.28.0-19-g98f9e67.98f9e67  161    latest/stable    canonical✓    -
gnome-3-34-1804          0+git.3556cb3               77     latest/stable/…  canonical✓    -
gnome-3-38-2004          0+git.1f9014a               99     latest/stable    canonical✓    -
gtk-common-themes        0.1-59-g7bca6ae             1519   latest/stable/…  canonical✓    -
kde-frameworks-5-core18  5.61.0                      32     latest/stable    kde✓          -
snap-store               3.38.0-66-gbd5b8f7          558    latest/stable/…  canonical✓    -
snapd                    2.54.4                      15177  latest/stable    canonical✓    snapd
```
and versions are:

```
dave@nuc3:~/snap$ code -v
1.65.2
c722ca6c7eed3d7987c0d5c3df5c45f6b15e77d1
x64
dave@nuc3:~/snap$ node -v
v16.14.0
```
The only differences I find between a standard bash terminal and a vscode terminal are with the paths:

VSCode terminal:
```
dave@nuc3:~/Shared/Projects/SevenSeas$ echo $PATH
/home/dave/dotnet:/home/dave/.netlify/helper/bin:/home/dave/.nvm/versions/node/v16.14.0/bin:/home/dave/dotnet:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/dave/Projects/flutter/bin:/home/dave/Projects/flutter/bin
```
Standard terminal
```
dave@nuc3:~/Shared/Projects/SevenSeas$ echo $PATH
/home/dave/.netlify/helper/bin:/home/dave/.nvm/versions/node/v16.14.0/bin:/home/dave/dotnet:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/dave/Projects/flutter/bin

```

**Comment by davegardner** [comment, marked as answer]
I cleaned up the $PATH and it all works fine. I really should have done that first :-(
